Fix #15735 Remove the Refresh modal class

Priyansh61 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #15735
2. This PR does the following: Fix the language change in exploration

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

https://user-images.githubusercontent.com/96513964/197282296-85a7de23-cbe3-40a8-a674-4fd6caa9d359.mp4


<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Priyansh61 commented:
@nithinrdy PTAL!

nithinrdy commented:
Hey, could you please explain the ideal expected behaviour (the behaviour that this PR is trying to achieve) just for my understanding?

Priyansh61 commented:
Thank you @nithinrdy for reviewing the PR.

Since a refresh is not required to change the language that appears to the user after selecting the language translation option, the PR eliminates the refresh functionality.

This does help to prevent unnecessary language switches to English when the user chooses another language.

In order to change the language without requiring a refresh, I removed the refresh modal class along with the tests and added some functionality.

I hope this helps :smile:

nithinrdy commented:
@Priyansh61 the issue description says that the issue arises when the user makes some progress (completes a state, for example), and then changes the language.

> Select an option using radio button which redirects to next screen (i.e. complete a card/state)
Now switch to spanish language which auto-reloads the page with popup


The video in the PR description doesn't display that user journey. The language is changed before even the first card is completed.

Could you please upload a new video where you first complete a state and _then_ change the language, and show that the issue no longer occurs?

Priyansh61 commented:
@nithinrdy I just had a doubt, we cant change the language of a card when we move to the next one since there is no language selection option available on the new card.

https://user-images.githubusercontent.com/96513964/198604181-ef6d17ff-cf09-4376-871f-cd660208d2e9.mp4

So is this what you were asking me to show ?? :sweat_smile:

nithinrdy commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.ts:
```diff
@@ -30,9 +30,6 @@ import { PlayerPositionService } from
   'pages/exploration-player-page/services/player-position.service';
 import { PlayerTranscriptService } from
   'pages/exploration-player-page/services/player-transcript.service';
-import { SwitchContentLanguageRefreshRequiredModalComponent } from
```
If this is the only place this component is being used, and we've now removed it, can we get rid of the component from the codebase?

Priyansh61 replied:
@nithinrdy Definitely!

I have checked for the component coverage and it is only used here to refresh the page when the user inputs something and selects the language change option.

Here's a video showing the changes when the component is completely removed from the codebase.

https://user-images.githubusercontent.com/96513964/198983250-f97d3065-2732-46e5-80db-f3bb7551b27d.mp4

nithinrdy reviewed (commented):
Thanks! Left a comment.

Priyansh61 commented:
@nithinrdy PTAL!

nithinrdy reviewed (approved):
@kevintab95 changes look good to me; could you please take a look?

gp201 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -144,29 +135,4 @@ describe('Content language selector component', () => {
     expect(imagePreloaderService.restartImagePreloader).toHaveBeenCalled();
   });
-
-  it('should correctly open the refresh required modal when refresh is ' +
```
Please rephrase the test above to reflect what is actually being tested.

gp201 reviewed (changes requested):
Thanks, @Priyansh61 ! I left some comments. PTAL

kevintab95 reviewed (changes requested):
Thanks @Priyansh61 for the PR. I dug into this a bit and here's what I found:

There are 2 ways to update the language of an exploration. The first way is to update the "global language" this will refresh the entire page and all of the text displayed on the screen will be of the selected global language. The second way is to just update the language of the exploration being played (e.g. the user decides that they want to view the page in English but wants only the content of the lesson to be in a different language -- selected by the dropdown in the exploration.)

Previously when we updated the content of an exploration (2nd method), it would add a URL param and refresh the page. When the page loads, it will look for this URL param and set the exploration player language to be that. We need to maintain this because if a user copies this URL, they can expect to view the exploration in the language they selected.

The issue seems to be that, we override the `currentContentLanguageCode` with the global one here:

https://github.com/oppia/oppia/blob/ef23203792be046abc6fbab569b645a36eb24706/core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.ts#L67-L76

So we might need to first clarify what needs to take precedence i.e. content language vs. global language.

@seanlip, any thoughts on this? What the behaviour should be in the following scenarios:
1. Site language is "en", the content is updated to "pt-br" and refreshed (with the relevant info in the URL). What should the user experience be? e.g. should we render an English page with pt-br lesson content?
2. Consider a scenario where the user views the page in pt-br (globally), then they decide to switch the lesson content to English, what should the user experience be?
3. User copies the URL of the lesson they are viewing and sends it to a friend (may or may not understand the same language), do we preserve the language details from the first user? What should the friends experience be when viewing the lesson (i.e. what should the lesson content and global language be)?

cc @dchen97

Priyansh61 commented:
@kevintab95 Regarding the aforementioned query, I had a few thoughts to add:

> Site language is "en", the content is updated to "pt-br" and refreshed (with the relevant info in the URL). What should the user experience be? e.g. should we render an English page with pt-br lesson content?

1. Creating an English page with lesson content in "pt-br" would be very helpful to the user.

> Consider a scenario where the user views the page in pt-br (globally), then they decide to switch the lesson content to English, what should the user experience be?

2. Since there is a good chance that the user may get stopped someplace and want to learn what it means in English, we should only convert the lesson material to English and leave the global language set to "pt-br" when the user wishes to switch from "pt-br" to English while exploring the card. We can simply switch the lesson's content language to English rather than doing a thorough refresh and changing the global language.

> User copies the URL of the lesson they are viewing and sends it to a friend (may or may not understand the same language), do we preserve the language details from the first user? What should the friends experience be when viewing the lesson (i.e. what should the lesson content and global language be)?

3. The url param, which loads the lesson in "pt-br," is included in the link when the user sends it to a friend. This is useful because, in most cases, the recipient of the link is a native speaker of the same region. Therefore page should be loaded in the local language, which will enhance the user experience.

cc @seanlip

seanlip commented:
Let's wait for @dchen97 (who leads PM for Web) to weigh in on this. She's already thought quite carefully about languages on the Web platform so I'd defer to her here.

dchen97 commented:
+1 to what @Priyansh61 said - language in lesson should respect the selected lesson language and the rest of the platform should respect the global language.

kevintab95 commented:
Thanks @dchen97! @Priyansh61, in that case, it sounds like we need to update the logic mentioned in https://github.com/oppia/oppia/pull/16378#pullrequestreview-1171558956 i.e. not override the lesson language.

Priyansh61 commented:
@kevintab95 Could you please tell me what changes I need to make before we proceed?

For example, should I include the refresh modal class and make some changes, and if so, what changes should be made? These things are a little unclear to me.

kevintab95 commented:
> +1 to what @Priyansh61 said - language in lesson should respect the selected lesson language and the rest of the platform should respect the global language.

Hi @Priyansh61, I think we will be keeping the "refresh modal" as-is but modifying the if-else mentioned in https://github.com/oppia/oppia/pull/16378#pullrequestreview-1171558956 such that we don't override the value of `selectedLanguageCode` if it is already set. Does that make sense?

Priyansh61 committed:
requirement.txt updated

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
added condition to check search param

Priyansh61 committed:
lint check fix

Priyansh61 committed:
lint check fix

Priyansh61 commented:
@kevintab95 Thanks a lot :+1: for the assistance. I have added the refresh modal again, with the condition that the language changes to global language only if there is no url search param and incase if it is present then the supplied language will be picked which is provided in the url.

Here's a proof for the above changes

https://user-images.githubusercontent.com/96513964/205281795-18baa391-d859-4dfa-8408-fac3e639da08.mp4

`

kevintab95 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.ts:
```diff
@@ -61,14 +63,22 @@ export class ContentLanguageSelectorComponent implements OnInit {
       this.contentTranslationLanguageService.getLanguageOptionsForDropdown());
     for (let option of this.languageOptions) {
-      if (option.value === this.currentGlobalLanguageCode) {
+      if (url.searchParams.get('initialContentLanguageCode') === option.value) {
```
I am a bit confused about the logic here, it seems like the if and else blocks execute the same code, is that intentional?

Priyansh61 replied:
@kevintab95 I have updated the logic so that we can avoid code duplication. PTAL and let me know for further changes.

kevintab95 reviewed (commented):
@Priyansh61 left a comment. PTAL

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
language update logic changed

Priyansh61 committed:
Merge branch 'language' of github.com:Priyansh61/oppia into language

Priyansh61 committed:
fixed lint checks

Priyansh61 committed:
lint check fix

Priyansh61 committed:
lint check fix

kevintab95 reviewed (approved):
LGTM, thanks @Priyansh61!

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
Merge branch 'develop' into language

gp201 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.ts:
```diff
@@ -51,24 +52,30 @@ export class ContentLanguageSelectorComponent implements OnInit {
       this.contentTranslationLanguageService.getCurrentContentLanguageCode());
     this.languageOptions = (
       this.contentTranslationLanguageService.getLanguageOptionsForDropdown());
+    this.newLanguageCode = (
+      url.searchParams.get('initialContentLanguageCode') ||
+      this.currentGlobalLanguageCode);
```
Please add a check for the value for this variable.

Priyansh61 replied:
@gp201 what do you mean by check? Do I need to write test for this variable ?

Priyansh61 replied:
Hi! @gp201 I tried to add the tests but since I'm new to the jasmine test , so if i have done something wrong please let me know and  I'd be glad it if you could give me some advice on how to properly write the test for this.

gp201 replied:
@Priyansh61 you wrote the `expect` stmt correctly. Can you combine it with the test above?

Priyansh61 replied:
@gp201 I am confused on how to get urll.searchParam variable in the test, like for the globalCurrentLanguageCode i have mocked the service in the test but I dont know how to do the same for url.searchParam since that includes mocking window service and many others.

Could you please help me fix the above problem ?

gp201 replied:
@Priyansh61 You are on the right path. You have to mock the window service. Use the file below for reference and try to solve it in a similar manner.
https://github.com/oppia/oppia/blob/af8dc516119a3198c33bf6947ce6c5f655e0e4a7/core/templates/pages/email-dashboard-pages/email-dashboard-result.component.spec.ts#L32-L40

Priyansh61 replied:
thanks @gp201 : )

gp201 reviewed (changes requested):
Thanks, @Priyansh61 ! I left some comments. PTAL

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
added check for newLanguageCode

gp201 reviewed (changes requested):
1 final comment

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
updated test

Priyansh61 committed:
Merge branch 'language' of github.com:Priyansh61/oppia into language

Priyansh61 committed:
Merge branch 'develop' into language

Shivkant-Chauhan commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -117,16 +117,17 @@ describe('Content language selector component', () => {
 
-  it('should correctly initialize selectedLanguageCode and ' +
-     'languagesInExploration', () => {
+  it('should correctly initialize selectedLanguageCode, ' +
+     'newLanguageCode and languagesInExploration', () => {
     expect(component.selectedLanguageCode).toBe('fr');
+    expect(component.newLanguageCode).toBe('fr');
     expect(component.languageOptions).toEqual([
       {value: 'fr', displayed: 'français (French)'},
       {value: 'zh', displayed: '中文 (Chinese)'},
       {value: 'en', displayed: 'English'}
     ]);
   });
```
```suggestion
it('should correctly initialize selectedLanguageCode, ' +
     'newLanguageCode and languagesInExploration', () => {
    expect(component.selectedLanguageCode).toBe('fr');
    expect(component.newLanguageCode).toBe('fr');
    expect(component.languageOptions).toEqual([
      {value: 'fr', displayed: 'français (French)'},
      {value: 'zh', displayed: '中文 (Chinese)'},
      {value: 'en', displayed: 'English'}
    ]);
  });

  it('should correcly initialize newLanguageCode', () => {
    let mockWindowRef = new MockWindowRef();
    let url = new URL(mockWindowRef.nativeWindow.location.href);

    expect(url.searchParams.get('initialContentLanguageCode')).toBe('fr');

    component.ngOnInit();

    expect(component.newLanguageCode).toBe('fr');
  });```

Shivkant-Chauhan commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -117,16 +117,17 @@ describe('Content language selector component', () => {
 
-  it('should correctly initialize selectedLanguageCode and ' +
-     'languagesInExploration', () => {
+  it('should correctly initialize selectedLanguageCode, ' +
+     'newLanguageCode and languagesInExploration', () => {
     expect(component.selectedLanguageCode).toBe('fr');
+    expect(component.newLanguageCode).toBe('fr');
     expect(component.languageOptions).toEqual([
       {value: 'fr', displayed: 'français (French)'},
       {value: 'zh', displayed: '中文 (Chinese)'},
       {value: 'en', displayed: 'English'}
     ]);
   });
```
```suggestion
  it('should correctly initialize selectedLanguageCode and , ' +
     'languagesInExploration', () => {
    expect(component.selectedLanguageCode).toBe('fr');
    expect(component.languageOptions).toEqual([
      {value: 'fr', displayed: 'français (French)'},
      {value: 'zh', displayed: '中文 (Chinese)'},
      {value: 'en', displayed: 'English'}
    ]);
  });

  it('should correcly initialize newLanguageCode', () => {
    let mockWindowRef = new MockWindowRef();
    let url = new URL(mockWindowRef.nativeWindow.location.href);

    expect(url.searchParams.get('initialContentLanguageCode')).toBe('fr');

    component.ngOnInit();

    expect(component.newLanguageCode).toBe('fr');
  });
```

Shivkant-Chauhan replied:
also add this class after line numb.70
```
class MockWindowRef {
  nativeWindow = {
    location: {
      href: 'http://localhost:8181/explore/wZiXFx1iV5bz?initialContentLanguageCode=fr'
    }
  };
}
```
this is simply mocking the WindowRef with it required properties

Priyansh61 replied:
Thanks @Shivkant-Chauhan for the help.

Priyansh61 committed:
test updated as per suggestions using windowRef for both the conditions

Priyansh61 committed:
Merge branch 'language' of github.com:Priyansh61/oppia into language

Priyansh61 committed:
fixed lint checks

Priyansh61 committed:
suggested commit

Priyansh61 committed:
Merge branch 'develop' into language

gp201 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -117,16 +125,24 @@ describe('Content language selector component', () => {
+    expect(url.searchParams.get('initialContentLanguageCode')).toBe('fr');
+    component.ngOnInit();
+    expect(component.newLanguageCode).toBe('fr');
+  });
```
```suggestion
  it('should correcly initialize newLanguageCode', () => {
    let mockWindowRef = new MockWindowRef();
    let url = new URL(mockWindowRef.nativeWindow.location.href);

    expect(url.searchParams.get('initialContentLanguageCode')).toBe('fr');

    component.ngOnInit();

    expect(component.newLanguageCode).toBe('fr');
  });
```

gp201 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -117,16 +125,24 @@ describe('Content language selector component', () => {
+    let mockWindowRef = new MockWindowRef();
+    let url = new URL(mockWindowRef.nativeWindow.location.href);
+    expect(url.searchParams.get('initialContentLanguageCode')).toBe('fr');
+    component.ngOnInit();
```
add a pre-check here for `newLanguageCode`. Basically, we're checking if the value changes after `component.ngOnInit();` is called.

Priyansh61 replied:
Thanks! @gp201 When i did that the value is still comming out to be 'fr' even if i change the url initial language code it is like we are not linked to windowRef we always get the same value.

Any suggesion ? I am really sorry to bother you but I dont have any clue how to move further.

gp201 replied:
@Priyansh61 you seem to have forgotten to direct the component to use the class that was defined.
https://github.com/oppia/oppia/blob/5d951d8e21e2f7872dcb26e8618952412efe4312/core/templates/services/user.service.spec.ts#L57

Take a look at this file above on how to mock `WindowRef`.

Priyansh61 replied:
Thanks @gp201 I tried to use the windowRef provider and MockWindow Ref as useClass the test for new language code passes succesfully for the first time :sweat_smile:  the url changes and the newlanguageCode initialised itself according to the url.

Then the problem the remaining 3 cases suddenly fails and gives me the following error.

````console
Content language selector component should correctly select an option when refresh is not needed FAILED
        TypeError: Cannot read properties of undefined (reading 'split')
            at ContextService.getPageContext (core/templates/combined-tests.spec.js:213435:57336)
            at ContextService.isInExplorationEditorPage (core/templates/combined-tests.spec.js:213435:71135)
            at ContentLanguageSelectorComponent.onSelectLanguage (core/templates/combined-tests.spec.js:148323:25192)
            at UserContext.<anonymous> (core/templates/combined-tests.spec.js:148226:19)
            at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:733524:30)
            at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:733009:43)
            at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:733523:36)
            at ./node_modules/zone.js/dist/zone.js.Zone.run (core/templates/combined-tests.spec.js:733281:47)
            at runInTestZone (core/templates/combined-tests.spec.js:732564:38)
            at UserContext.<anonymous> (core/templates/combined-tests.spec.js:732579:24)
````


Can you please help me with this ? The only change I made is 

```typescript
{
        provide: WindowRef,
       useClass: MockWindowRef
},
```

The above change makes the new test case pass but fails all the other tests

gp201 replied:
The error given below:
```console
TypeError: Cannot read properties of undefined (reading 'split')
```

tells that there is an undefined value. Can you check if a variable that is used in other tests hasn't been defined?

Priyansh61 replied:
Thank you very much, @gp201, for guiding me. Actually, one of the tests required the windowRef's pathname attribute as well. I included it now, and all of the tests are finally passing. I also verified for both scenarios when searchPram is present and when it is not there, and if present, it returns en and otherwise fr.

gp201 reviewed (changes requested):
Good job on mocking windowref. A few more nits.

Priyansh61 committed:
Merge branch 'develop' into language

gp201 commented:
@Priyansh61 I replied to your comment.

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
tests fixed

Priyansh61 committed:
lint check fixed

Priyansh61 committed:
lint checks fixed

Priyansh61 committed:
Merge branch 'develop' into language

gp201 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -117,16 +131,30 @@ describe('Content language selector component', () => {
   });
 
+  it('should correcly initialize newLanguageCode', () => {
+    windowRef = TestBed.inject(WindowRef);
```
Can you move this line to the `beforeEach` block?

gp201 commented on core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts:
```diff
@@ -117,16 +131,30 @@ describe('Content language selector component', () => {
+
+    windowRef.nativeWindow.location.href = (
+      'http://localhost:8181/explore/wZiXFx1iV5bz?initialContentLanguageCode=en');
+    windowRef.nativeWindow.location.pathname = (
+      '/explore/wZiXFx1iV5bz?initialContentLanguageCode=en');
+    component.ngOnInit();
+    expect(component.newLanguageCode).toBe('en');
```
```suggestion
    windowRef.nativeWindow.location.pathname = (
      '/explore/wZiXFx1iV5bz?initialContentLanguageCode=en');
 
    component.ngOnInit();

    expect(component.newLanguageCode).toBe('en');
```

gp201 reviewed (approved):
@Priyansh61 great job on the tests. Only some lint issues left. PTAL

gp201 reviewed (changes requested):
I left a review above PTAL

Priyansh61 committed:
Update core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts

Co-authored-by: Praneeth <praneethg2001@gmail.com>

Priyansh61 committed:
moved the inject statement

Priyansh61 committed:
lint checks

Priyansh61 committed:
Merge branch 'develop' into language

Priyansh61 committed:
lint fix

Priyansh61 committed:
Merge branch 'language' of github.com:Priyansh61/oppia into language

Priyansh61 commented:
@gp201 PTAL! I have addressed all the requested changes :)

gp201 reviewed (approved):
LGTM
